zokrates_core/src/proof_system/ark/groth16.rs

```diff
@@ -0,0 +1,214 @@
+        program: ProgIterator<T, I>,
+        witness: Witness<T>,
+        proving_key: Vec<u8>,
+    ) -> Proof<<G16 as Scheme<T>>::ProofPoints> {
+        println!("{}", G16_WARNING);
+
+        let computation = Computation::with_witness(program, witness);
+
+        let inputs = computation
+            .public_inputs_values()
+            .iter()
+            .map(parse_fr::<T>)
+            .collect::<Vec<_>>();
+
+        let pk = ProvingKey::<<T as ArkFieldExtensions>::ArkEngine>::deserialize_uncompressed(
+            &mut proving_key.as_slice(),
+        )
+        .unwrap();
+
+        let rng = &mut rand_0_7::rngs::StdRng::from_seed(get_random_seed().unwrap());
+        let proof = Groth16::<T::ArkEngine>::prove(&pk, computation, rng).unwrap();
+
+        let proof_points = ProofPoints {
+            a: parse_g1::<T>(&proof.a),
+            b: parse_g2::<T>(&proof.b),
+            c: parse_g1::<T>(&proof.c),
+        };
+
+        Proof::new(proof_points, inputs)
+    }
+
+    fn verify(
+        vk: <G16 as Scheme<T>>::VerificationKey,
+        proof: Proof<<G16 as Scheme<T>>::ProofPoints>,
+    ) -> bool {
+        let vk = VerifyingKey {
+            alpha_g1: serialization::to_g1::<T>(vk.alpha),
+            beta_g2: serialization::to_g2::<T>(vk.beta),
+            gamma_g2: serialization::to_g2::<T>(vk.gamma),
+            delta_g2: serialization::to_g2::<T>(vk.delta),
+            gamma_abc_g1: vk
+                .gamma_abc
+                .into_iter()
+                .map(serialization::to_g1::<T>)
+                .collect(),
+        };
+
+        let pvk: PreparedVerifyingKey<T::ArkEngine> = prepare_verifying_key(&vk);
+        let ark_proof = ArkProof {
+            a: serialization::to_g1::<T>(proof.proof.a),
+            b: serialization::to_g2::<T>(proof.proof.b),
+            c: serialization::to_g1::<T>(proof.proof.c),
+        };
+
+        let public_inputs: Vec<_> = proof
+            .inputs
+            .iter()
+            .map(|s| {
+                T::try_from_str(s.trim_start_matches("0x"), 16)
+                    .unwrap()
+                    .into_ark()
+            })
+            .collect::<Vec<_>>();
+
+        verify_proof(&pvk, &ark_proof, &public_inputs).unwrap()
+    }
+}
+
+impl<T: Field + ArkFieldExtensions + NotBw6_761Field> NonUniversalBackend<T, G16> for Ark {
+    fn setup<I: IntoIterator<Item = Statement<T>>>(
+        program: ProgIterator<T, I>,
+    ) -> SetupKeypair<<G16 as Scheme<T>>::VerificationKey> {
+        println!("{}", G16_WARNING);
+
+        let computation = Computation::without_witness(program);
+
+        let rng = &mut rand_0_7::rngs::StdRng::from_seed(get_random_seed().unwrap());
+        let (pk, vk) = Groth16::<T::ArkEngine>::circuit_specific_setup(computation, rng).unwrap();
+
+        let mut pk_vec: Vec<u8> = Vec::new();
+        pk.serialize_uncompressed(&mut pk_vec).unwrap();
+
+        let vk = VerificationKey {
```

Comment:
There is a lot of code constructing these vks and proofs and it's used many times, is there a way to implement the From trait once and reuse it instead?

Reply:
This will be done in another PR